src/drivers/intel/ssp/ssp.c

```diff
@@ -125,6 +125,111 @@ static int ssp_context_restore(struct dai *dai)
 	return 0;
 }
 
+static int ssp_mclk_prepare_enable(struct dai *dai)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	struct sof_ipc_dai_config *config = &ssp->config;
+	int ret;
+
+	if (ssp->clk_active & SSP_CLK_MCLK_ACTIVE)
+		return 0;
+
+	/* MCLK config */
+	ret = mn_set_mclk(config->ssp.mclk_id, config->ssp.mclk_rate);
+	if (ret < 0)
+		dai_err(dai, "ssp_mclk_prepare_enable(): invalid mclk_rate = %d for mclk_id = %d",
+			config->ssp.mclk_rate, config->ssp.mclk_id);
+	else
+		ssp->clk_active |= SSP_CLK_MCLK_ACTIVE;
+
+	return ret;
+}
+
+static void ssp_mclk_disable_unprepare(struct dai *dai)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+
+	if (!(ssp->clk_active & SSP_CLK_MCLK_ACTIVE))
+		return;
```

Comment:
does this work if both playback and capture on the same SSP active simultaneously, it is released at the stop of the first stream no?

Reply:
> does this work if both playback and capture on the same SSP active simultaneously, it is released at the stop of the first stream no?

We check ssp->state in both `ssp_set_config` and `ssp_post_stop`. So it should be fine. i.e. `ssp_mclk_disable_unprepare` wont be called when any of ssp->state is ACTIVE.